src/core/primitives.rs

```diff
@@ -451,3 +431,63 @@ fn refresh_coordinate_keys(
         })
         .collect::<RevisionVec<_, _>>()
 }
+
+/// Refreshes the encapsulation with the given MSK.
+pub fn reencrypt(
+    rng: &mut impl CryptoRngCore,
+    msk: &MasterSecretKey,
+    mpk: &MasterPublicKey,
+    encapsulation: &XEnc,
+) -> Result<(Secret<32>, XEnc), Error> {
+    let (_secret, rights) = full_decaps(msk, encapsulation)?;
+    let vec_to_hashset = HashSet::from_iter(rights);
+    let new_encaps = encaps(rng, mpk, &vec_to_hashset)?;
+    Ok(new_encaps)
+}
+
+/// Attempts opening the Covercrypt encapsulation using the given MSK. Returns
+/// the encapsulated key and the list of given rights.
+fn full_decaps(
+    msk: &MasterSecretKey,
+    encapsulation: &XEnc,
+) -> Result<(Secret<SHARED_SECRET_LENGTH>, Vec<Right>), Error> {
+    let A = encapsulation.c.first().unwrap() * &(&msk.tsk.s / &msk.tsk.tracers.front().unwrap().0);
+    let mut right_to_secrets = Vec::with_capacity(encapsulation.encapsulations.len());
```

Comment:
This is not `right_to_secret` anymore.
```suggestion
    let mut rights = Vec::with_capacity(encapsulation.encapsulations.len());
```